Add serde traits to Auth

RCasatta opened the pull request:
This would allow to ser/des `RpcConfig` downstream struct

sgeisler commented on client/src/client.rs:
```diff
@@ -187,7 +187,7 @@ impl RawTx for String {
 /// The different authentication methods for the client.
-#[derive(Clone, Debug, Hash, Eq, PartialEq, Ord, PartialOrd)]
+#[derive(Clone, Debug, Hash, Eq, PartialEq, Ord, PartialOrd, Serialize, Deserialize)]
 pub enum Auth {
```
I wonder if it would be worth it implementing the serde logic manually because none of the available tagging options feel right. The default, external tagging, would result in:

```json
"auth_field": {
  "UserPass": ["user", "pass"]
}
```

or
 
```json
"auth_field": {
  "CookieFile": "/path/to/file"
}
```

I think we should at least add variant renaming to snake_case. But ideally it would look like this imo:

```json
"auth_field": {
  "user": "user",
  "pass": "pass"
}
```

or
 
```json
"auth_field": {
  "cookie_file": "/path/to/file"
}
```

We could achieve this with untagged serialization and changing the tuple variants to struct variants, but that would be an API break and would make direct usage more cumbersome. We could add constructors to ease the pain, but the API break would remain.

sgeisler reviewed (commented):
Concept ACK, but we should discuss the representation.

RCasatta commented:
@sgeisler I updated the PR to serialize in a more comprehensible way and at the same time trying to be non-breaking and easily upgradable

we may want to add `#[serde(untagged)]` but I slightly prefer the default

sgeisler reviewed (commented):
Without `#[serde(untagged)]` it still seems unnecessarily deeply nested :confused: I think with `#[serde(untagged)]` and the `CookieFile` variant becoming a named struct instead of a tuple struct that might be quite nice. If you feel strongly about it I'm ok with this too though.

Also, if we were feeling really fancy we'd avoid the clone on serializing by having two structs: one owned and one borrowed.

RCasatta commented:
> Without #[serde(untagged)] it still seems unnecessarily deeply nested 😕 I think with #[serde(untagged)] and the CookieFile variant becoming a named struct instead of a tuple struct that might be quite nice. If you feel strongly about it I'm ok with this too though.

I am not feeling strongly about it and I added #[serde(untagged)] (but I wouldn't say "unnecessarily deeply nested" since there are reasons to prefer it...)
I also added the named variant for `Cookie`

> Also, if we were feeling really fancy we'd avoid the clone on serializing by having two structs: one owned and one borrowed.

I don't think the code complexity is worth it in this case since serialize/deserialize of this value is done rarely

sgeisler reviewed (approved):
utACK 4512d5ddb8cb2536e2a8cdd84266414bb9908126

> but I wouldn't say "unnecessarily deeply nested" since there are reasons to prefer it...

Sorry for that wording, I just didn't understand it :smile: I guess there are some problems if a parent enum is untagged too and you have collisions in field names (just don't do that :laughing: ). But since this is to be used in config (I suppose) and I as a user tend to get annoyed when I have to write seemingly redundant information I think this version is better.

> I don't think the code complexity is worth it in this case since serialize/deserialize of this value is done rarely

Very good point.

stevenroose commented:
Sorry for the delay guys.. It's a bit unfortunate that we have to clone stuff for this. I don't think it's too hard to implement serde manually without the need for an extra struct to copy into.

Even though the only problem with just doing this on our existing struct is that user and pass are a slice. Which is also not the end of the world:
```rust

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
enum Auth {
    None,
    UserPass(String, String),
    CookieFile(PathBuf),
}
```
which gives
```json
{"auth":"none"}
{"auth":{"user_pass":["userr","passs"]}}
{"auth":{"cookie_file":"/"}}
```

stevenroose commented:
Just for curiosity, I tried it manually. It ends up being quite verbose like always using the visitor etc. But it works without re-allocations for serializing: https://github.com/stevenroose/rust-bitcoincore-rpc/commit/04dd65b54e3fcfa049afa6b8c644cbb50646331b

I'm fine also just merging this if that's preferred.

RCasatta commented:
I am fine also with stevenroose@04dd65b
